Mathlib/Probability/Distributions/SetBernoulli.lean

```diff
@@ -0,0 +1,126 @@
+
+open MeasureTheory Measure unitInterval
+open scoped Finset
+
+namespace ProbabilityTheory
+variable {ι Ω : Type*} {m : MeasurableSpace Ω} {X Y : Ω → Set ι} {s u : Set ι} {i j : ι} {p q : I}
+  {P : Measure Ω}
+
+variable (u p) in
+/-- The product of bernoulli distributions with parameter `p` on the set `u : Set V` is the measure
+on `Set V` such that each element of `u` is taken with probability `p`, and the elements outside of
+`u` are never taken. -/
+@[expose]
+noncomputable def setBernoulli : Measure (Set ι) :=
+  .comap (fun s i ↦ i ∈ s) <| infinitePi fun i ↦
+    toNNReal p • dirac (i ∈ u) + toNNReal (σ p) • dirac False
+
+@[inherit_doc] scoped notation "setBer(" u ", " p ")" => setBernoulli u p
+
+instance : IsProbabilityMeasure setBer(u, p) :=
+  MeasurableEquiv.setOf.symm.measurableEmbedding.isProbabilityMeasure_comap <| .of_forall fun P ↦
+    ⟨{a | P a}, rfl⟩
+
+variable (u p) in
+lemma setBernoulli_eq_map :
+    setBer(u, p) = .map (fun p ↦ {i | p i})
+      (infinitePi fun i ↦ toNNReal p • dirac (i ∈ u) + toNNReal (σ p) • dirac False) :=
+  MeasurableEquiv.setOf.comap_symm
+
+lemma setBernoulli_apply (S : Set (Set ι)) :
+    setBer(u, p) S = (infinitePi fun a ↦ toNNReal p • dirac (a ∈ u) + toNNReal (σ p) • dirac False)
+      ((fun t a ↦ a ∈ t) '' S) := by
+  convert MeasurableEquiv.setOf.symm.measurableEmbedding.comap_apply ..
+
+lemma setBernoulli_apply' (S : Set (Set ι)) :
+    setBer(u, p) S = (infinitePi fun a ↦ toNNReal p • dirac (a ∈ u) + toNNReal (σ p) • dirac False)
+      ((fun p ↦ {a | p a}) ⁻¹' S) := by
+  convert MeasurableEquiv.setOf.symm.comap_apply ..
+
+variable (u) in
+@[simp] lemma setBernoulli_zero : setBer(u, 0) = dirac ∅ := by simp [setBernoulli_eq_map]
+
+variable (u) in
+@[simp] lemma setBernoulli_one : setBer(u, 1) = dirac u := by simp [setBernoulli_eq_map]
+
+section Countable
+variable [Countable ι]
+
+lemma setBernoulli_ae_subset : ∀ᵐ s ∂setBer(u, p), s ⊆ u := by
+  classical
+  change _ = _
+  simp only [Set.compl_setOf, Set.not_subset_iff_exists_mem_notMem, Set.setOf_exists, Set.setOf_and,
+    measure_iUnion_null_iff]
+  rintro a
+  by_cases ha : a ∈ u
+  · simp [*]
+  calc
+    setBer(u, p) ({s | a ∈ s} ∩ {s | a ∉ u})
+    _ = setBer(u, p) {s | a ∈ s} := by simp [ha]
+    _ = infinitePi (fun a ↦ toNNReal p • dirac (a ∈ u) + toNNReal (σ p) • dirac False)
+            (cylinder {a} {fun _ ↦ True}) := by
+      rw [setBernoulli_apply']
+      congr!
+      ext
+      simp [funext_iff]
+    _ = 0 := by simp [infinitePi_cylinder _ (.singleton _), ha]
+
+variable (u p) in
+@[simp] lemma setBernoulli_singleton (hsu : s ⊆ u) (hu : u.Finite) :
+    setBer(u, p) {s} = toNNReal p ^ s.ncard * toNNReal (σ p) ^ (u \ s).ncard := by
+  classical
+  lift u to Finset ι using hu
+  calc
+    setBer(u, p) {s}
+    _ = ∏' i, ((if i ∈ u ↔ i ∈ s then (toNNReal p : ENNReal) else 0) +
+          if i ∈ s then 0 else (toNNReal (σ p) : ENNReal)) := by
+      simp [setBernoulli_apply, Set.image_singleton, Set.indicator]
```

Comment:
nonterminal simp?

Reply:
If you are looking for a real improvement instead, there should be a better way to convert from `I` to `ENNReal`